specs/advanced-data-layouts/prollytree/spec.md

```diff
@@ -0,0 +1,508 @@
+Then, check if a removed leaf/node was a 'splitting node' and the nodes need to be merged (only the last node in a chunk can be a splitting node). If splitting criterion holds for the last item of the node and there is the current node is succeeded by another node on the same level: merge the current node with the succeeding node. Continue walking up the path. If at root and it is being split: Create a new root that links to the freshly split nodes. Return the new root CID.
+
+Pay attention to the fact that the boundary algorithm is not related to the node CID, i.e the boundary is generated *before* the node CID is generated.
+
+### Put/Remove/Update
+
+First search if the tree already contains the value.
+If the key is not contained, create a new leaf for the value, insert the key-value (key-leaf) pair in the correct node and rebalance the tree.
+If the key is already stored in the tree and the value is the same, the tree is left unchanged.
+If the key exists and the values differ: Either leave the tree or insert the new value and update the tree.
+If removing the key, remove the `key` and the `value` from the list. After making any modifications to the tree, run the (Rebalance)[#Rebalance] algorithm.
+
+## Structure
+
+```
+type ProllyTree struct {
+  config &ProllyTreeConfig
+  root &TreeNode
+} representation tuple
+
+type TreeNode struct {
+  isLeaf Boolean
+  keys [Bytes]
+  # If a leaf, contains entry valies
+  # If an intermediate node, contains Links to further TreeNodes
+  values [Any]
+} representation tuple
+
+type HashThresholdConfig struct {
+  chunkingFactor Int
+  hashFunction Int
+} representation tuple
+
+type ChunkingStrategy union {
+  | HashThresholdConfig "hashThreshold"
+} representation keyed
+
+type ProllyTreeConfig struct {
+  minChunkSize Int
+  maxChunkSize Int
+  cidVersion Int
+  codec Int
+  hashFunction Int
+  nullable hashLength Int
+  strategy ChunkingStrategy
+} representation tuple
+```
+
+### `ProllyTree`
+
+A prolly tree is identified by a node which links to the `root` `TreeNode` for the actual key-value pairs,
+and a `config` link to the ProllyTreeConfig which has infomration about the chunking and encoding information in order to quickly compare if two trees are using the same configuration and may be merged together as well as making it easier to write.
+
+### `TreeNode`
+
+This is the "Tree" part of Prolly Trees and is made to be general purpose.
+We can potentially expect to use the TreeNode structure in subsequent specs with different types of trees.
+
+### `TreeNode.keys`
+
+Raw keys(keys/values input from users) for leaf node. Key-value pairs are sorted by byte value with the "larger" keys being at the end. Values are compared at the first byte, and going down to the end. This means that keys that are just a prefix come before keys that are prefix + 1 byte.
+
+### `TreeNode.values`
+
```

Comment:
Not sure about what is said here.

"Values corresponding to keys."
"Values can point to arbitrary IPLD nodes and it is up to applications to generate and process them."
"For non-leaf node these will be Links to other nodes."

Would make more sense IMO.